### PR TITLE
Add the abstractions for SLVS' side of the SSE handling

### PR DESCRIPTION
**Pull request by georgii-borovinskikh-sonarsource**

### CODE REVIEW

**Comment by rita-sonarsource** [comment]
Create a jira ticket + link to it here

File: .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml [hunk 1]
```diff
@@ -44,7 +44,7 @@
           <Key>headerFormat</Key>
           <Value>/*
  * SonarLint for Visual Studio
- * Copyright (C) 2016-2022 SonarSource SA
+ * Copyright (C) 2016-2023 SonarSource SA
```
Comment (rita-sonarsource) [inline .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml:47]: Please do it in a separate PR

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSource.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.TEMP;
+
+namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
+{
+    internal interface IServerSentEventSource<T> where T : IServerSentEvent
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/IServerSentEventSource.cs:26]: Add a doc comment to all interfaces

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.TEMP;
+
+namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
+{
+    internal interface IServerSentEventSourcePublisher<T> where T : IServerSentEvent
+    {
+        Task Publish(T serverEvent);
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs:28]: doc comment what it returns

File: src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.TEMP;
+
+namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
+{
+    internal interface ITaintServerEventSource : IServerSentEventSource<ITaintServerEvent>{}
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs:25]: WDYT about putting the taint and issue under separate folders?
so
| ServerSentEvents
     | TaintVulnerabilities
     | Issues
     IServerSentEventSource.cs
     ...

File: src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using SonarLint.VisualStudio.Core.Binding;
+using SonarQube.Client;
+
+namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
+{
+    internal interface IServerSentEventSessionManager
+    {
+    }
+
+    [Export(typeof(IServerSentEventSessionManager))]
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs:31]: Singleton is the default behavior but we still make it explicit, `[PartCreationPolicy(CreationPolicy.Shared)]`

File: src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    internal class ServerSentEventSessionManager : IServerSentEventSessionManager
+    {
+        [ImportingConstructor]
+        public ServerSentEventSessionManager(
```
Comment (rita-sonarsource) [inline src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs:35]: Please add a MEF export test. Search for "MefCtor_CheckIsExported" for examples

**Review by rita-sonarsource — CHANGES_REQUESTED** [review]
See my comments

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs (hunk 3 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs:28]: Describe why it's a Task and not a void?

File: .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml (hunk 1 above)
Comment (georgii-borovinskikh-sonarsource) [reply .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml:47]: I'm not sure how to revert this, could you help?

File: src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs (hunk 4 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs:25]: Sure

**Comment by duncanp-sonar** [comment]
Don't mix unrelated changes in a single PR - the binding update is completely separate and should have its own commit/PR.

**Review by duncanp-sonar — APPROVED** [review]
Nothing to add to Rita's comments.

File: .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml (hunk 1 above)
Comment (rita-sonarsource) [reply .sonarlint/sonarlint-visualstudio/CSharp/SonarLint.xml]: All good 👍

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs (hunk 3 above)
Comment (rita-sonarsource) [reply src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs:28]: Yes (should probably be PublishAsync). i.e. can it throw?

File: src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs (hunk 5 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs:33]: done

File: src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs (hunk 6 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs:38]: done

File: src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs (hunk 4 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/ITaintServerEventSource.cs]: done

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs (hunk 3 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/IServerSentEventSourcePublisher.cs]: done

File: src/ConnectedMode/ServerSentEvents/IServerSentEventSource.cs (hunk 2 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/ServerSentEvents/IServerSentEventSource.cs:30]: done

**Review by rita-sonarsource — APPROVED** [review]
LGTM
